4-functional-programming/src/main.ts

```diff
@@ -1,22 +1,74 @@
-import { Either, fromPromise, ap, right, getOrElse, flatten } from './fp/either';
-import { pipe } from './fp/utils';
+import { Either, fromPromise, ap, right, flatten, left, getOrElse } from './fp/either';
+import { flow, matcher, pipe, prop } from './fp/utils';
 import { fetchClient, fetchExecutor } from './fetching';
-import { ClientUser, ExecutorUser } from './types';
+import { ClientUser, Demand, ExecutorUser } from './types';
+import { fold, fromNullable, isNone } from './fp/maybe';
+import { map, sort } from './fp/array';
+import { distance } from './utils';
+import { fromCompare, ordNumber } from './fp/ord';
 
 type Response<R> = Promise<Either<string, R>>
 
 const getExecutor = (): Response<ExecutorUser> => fromPromise(fetchExecutor());
-const getClients = (): Response<Array<ClientUser>> => fromPromise(fetchClient());
+const getClients = (): Response<Array<ClientUser>> => fromPromise(fetchClient().then(map(client => ({
+  ...client,
+  demands: fromNullable(client.demands),
+}))));
 
 export enum SortBy {
   distance = 'distance',
   reward = 'reward',
 }
 
 export const show = (sortBy: SortBy) => (clients: Array<ClientUser>) => (executor: ExecutorUser): Either<string, string> => {
+  interface MappedClient extends ClientUser {
+    distance: number,
+  }
+  
+  const mappedClients: MappedClient[] = clients
+    .map((client) => ({
+      ...client,
+      distance: distance(executor.position, client.position),
+    }));
 
+  const ord = fromCompare((client1: MappedClient, client2: MappedClient) => (
+    sortBy === 'reward'
```

Comment:
```suggestion
    sortBy === SortBy.reward
```
Better than just a string.

Reply:
Fixed